File: render.c
#include <stddef.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"

#define IDX(x, y, width) (y * width + x)
#define PI 3.14159265358979323846

typedef struct {
    int r;
    int g;
    int b;
} Color;

typedef struct {
    float x;
    float y;
    float z;
} Vec3;

typedef struct {
    float dist;
    Vec3 normal;
} RayHit;

typedef struct {
    Vec3* vertices;
    int num_vertices;
    int num_faces;
    int* faces;
    Color* colors;
} Object;

typedef struct {
    Vec3 forward;
    Vec3 right;
    Vec3 up;
} Camera;



Vec3 vec3_add(Vec3 a, Vec3 b) {
    return (Vec3) {a.x + b.x, a.y + b.y, a.z + b.z};
}

Vec3 vec3_sub(Vec3 a, Vec3 b) {
    return (Vec3) {a.x - b.x, a.y - b.y, a.z - b.z};
}

Vec3 vec3_scale(Vec3 a, double scalar) {
    return (Vec3) {a.x * scalar, a.y * scalar, a.z * scalar};
}

float vec3_dot(Vec3 a, Vec3 b) {
    return a.x * b.x + a.y * b.y + a.z * b.z;
}

Vec3 vec3_cross(Vec3 a, Vec3 b) {
    return (Vec3) {a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x};
}

float vec3_length(Vec3 a) {
    return sqrt(a.x * a.x + a.y * a.y + a.z * a.z);
}

Vec3 vec3_normalize(Vec3 a) {
    float length = vec3_length(a);
    return (Vec3) {a.x / length, a.y / length, a.z / length};
}

void vec3_print(Vec3 a) {
    printf("(%f, %f, %f)\n", a.x, a.y, a.z);
}

RayHit* ray_triangle_intersection(Vec3 origin, Vec3 direction, Vec3 v0, Vec3 v1, Vec3 v2) {
    Vec3 e1 = vec3_sub(v1, v0);
    Vec3 e2 = vec3_sub(v2, v0);
    Vec3 p = vec3_cross(direction, e2);
    float det = vec3_dot(e1, p);
    if (det < 1e-6) {
        return NULL;
    }
    float inv_det = 1.0 / det;
    Vec3 t = vec3_sub(origin, v0);
    float u = vec3_dot(t, p) * inv_det;
    if (u < 0.0 || u > 1.0) {
        return NULL;
    }
    Vec3 q = vec3_cross(t, e1);
    float v = vec3_dot(direction, q) * inv_det;
    if (v < 0.0 || u + v > 1.0) {
        return NULL;
    }
    float dist = vec3_dot(e2, q) * inv_det;
    Vec3 normal = vec3_cross(e1, e2);
    RayHit* hit = (RayHit*) malloc(sizeof(RayHit));
    hit->dist = dist;
    hit->normal = vec3_normalize(normal);
    return hit;
}

Camera* calculate_camera_basis(Vec3 direction) {
    Vec3 forward = vec3_normalize(direction);
    Vec3 worldUp = (Vec3) {0.0, 1.0, 0.0};
    Vec3 right = vec3_cross(worldUp, forward);
    if (vec3_length(right) < 1e-6) {
        worldUp = (Vec3) {1.0, 0.0, 0.0};
        right = vec3_cross(worldUp, forward);
    }
    right = vec3_normalize(right);
    Vec3 up = vec3_cross(forward, right);
    up = vec3_normalize(up);
    up = vec3_scale(up, -1.0);
    right = vec3_scale(right, -1.0);
    Camera* camera = (Camera*) malloc(sizeof(Camera));
    camera->forward = forward;
    camera->right = right;
    camera->up = up;
    return camera;
}


unsigned char* render(int width, int height, Object** objects, int num_objects, Vec3 origin, Vec3 direction, float fov) {
    Vec3 lightDir = (Vec3) {0.0, 0.0, 1.0};
    unsigned char* image = (unsigned char*) malloc(3 * width * height * sizeof(unsigned char));
    float* depthMap = (float*) malloc(width * height * sizeof(float));

    Camera* camera = calculate_camera_basis(direction);
    Vec3 forward = camera->forward;
    Vec3 right = camera->right;
    Vec3 up = camera->up;

    float aspect_ratio = (float) width / height;
    float fov_rad = fov * (float) PI / 180.0;
    float half_tan_fov = (float) tan(fov_rad * 0.5);

    for (int i = 0; i < width; i++) {
        for (int j = 0; j < height; j++) {
            float x = (2.0 * (i + 0.5) / width - 1.0) * aspect_ratio * half_tan_fov;
            float y = (2.0 * (j + 0.5) / height - 1.0) * half_tan_fov;
            Vec3 ray_direction = vec3_add(forward, vec3_add(vec3_scale(right, x), vec3_scale(up, y)));
            ray_direction = vec3_normalize(ray_direction);

            float closest_t = INFINITY;
            Color closest_color = (Color) {0, 0, 0};

            for (int object_index = 0; object_index < num_objects; object_index++) {
                Object* object = objects[object_index];

                for (int face_index = 0; face_index < object->num_faces; face_index++) {
                    int base_idx = face_index * 3;
                    int idx0 = object->faces[base_idx];
                    int idx1 = object->faces[base_idx + 1];
                    int idx2 = object->faces[base_idx + 2];
                    Vec3 a = object->vertices[idx0];
                    Vec3 b = object->vertices[idx1];
                    Vec3 c = object->vertices[idx2];
                    RayHit* hit = ray_triangle_intersection(origin, ray_direction, a, b, c);
                    if (hit != NULL && hit->dist > 0.0 && hit->dist < closest_t) {
                        closest_t = hit->dist;
                        Color color = object->colors[face_index];
                        double d = fmax(0.4, vec3_dot(hit->normal, lightDir));
                        closest_color = (Color) {
                            (int) (color.r * d),
                            (int) (color.g * d),
                            (int) (color.b * d)
                        };
                        free(hit);
                    }
                }
            }

            int idx = IDX(i, j, width);
            image[3*idx] = closest_color.r;
            image[3*idx + 1] = closest_color.g;
            image[3*idx + 2] = closest_color.b;
            depthMap[idx] = closest_t;
        }
    }

    free(camera);
    free(depthMap);
    return image;
}

Object* create_debug_cube(float size) {
    float half_size = size / 2.0;
    Vec3* vertices = (Vec3*) malloc(8 * sizeof(Vec3));
    vertices[0] = (Vec3) {-half_size, -half_size, -half_size};
    vertices[1] = (Vec3) { half_size, -half_size, -half_size};
    vertices[2] = (Vec3) { half_size,  half_size, -half_size};
    vertices[3] = (Vec3) {-half_size,  half_size, -half_size};
    vertices[4] = (Vec3) {-half_size, -half_size,  half_size};
    vertices[5] = (Vec3) { half_size, -half_size,  half_size};
    vertices[6] = (Vec3) { half_size,  half_size,  half_size};
    vertices[7] = (Vec3) {-half_size,  half_size,  half_size};

    int* faces = (int*) malloc(3 * 12 * sizeof(int));  // 12 faces * 3 vertices per face
    // Face 0: front face
    faces[0] = 0; faces[1] = 2; faces[2] = 1;
    faces[3] = 0; faces[4] = 3; faces[5] = 2;
    // Face 1: back face  
    faces[6] = 4; faces[7] = 5; faces[8] = 6;
    faces[9] = 4; faces[10] = 6; faces[11] = 7;
    // Face 2: left face
    faces[12] = 0; faces[13] = 4; faces[14] = 7;
    faces[15] = 0; faces[16] = 7; faces[17] = 3;
    // Face 3: right face
    faces[18] = 1; faces[19] = 2; faces[20] = 6;
    faces[21] = 1; faces[22] = 6; faces[23] = 5;
    // Face 4: bottom face
    faces[24] = 0; faces[25] = 1; faces[26] = 5;
    faces[27] = 0; faces[28] = 5; faces[29] = 4;
    // Face 5: top face
    faces[30] = 2; faces[31] = 3; faces[32] = 7;
    faces[33] = 2; faces[34] = 7; faces[35] = 6;

    Color* colors = (Color*) malloc(12 * sizeof(Color));
    colors[0] = (Color) {0, 0, 255};
    colors[1] = (Color) {0, 0, 255};
    colors[2] = (Color) {0, 255, 0};
    colors[3] = (Color) {0, 255, 0};
    colors[4] = (Color) {255, 0, 0};
    colors[5] = (Color) {255, 0, 0};
    colors[6] = (Color) {255, 255, 0};
    colors[7] = (Color) {255, 255, 0};
    colors[8] = (Color) {255, 0, 255};
    colors[9] = (Color) {255, 0, 255};
    colors[10] = (Color) {0, 255, 255};
    colors[11] = (Color) {0, 255, 255};
    
    Object* object = (Object*) malloc(sizeof(Object));
    object->vertices = vertices;
    object->num_faces = 12;
    object->num_vertices = 8;
    object->faces = faces;
    object->colors = colors;
    return object;
}

Object* read_obj(char* filename, Color color) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error: could not open file %s\n", filename);
        return NULL;
    }
    Vec3* vertices = (Vec3*) malloc(sizeof(Vec3));
    int num_vertices = 0;
    int vertex_capacity = 1;
    int* faces = (int*) malloc(3*sizeof(int));
    int num_faces = 0;
    int face_capacity = 1;

    char line[1024];
    while (fgets(line, sizeof(line), file)) {
        if (line[0] == 'v') {
            int pos = num_vertices;
            num_vertices++;
            if (num_vertices > vertex_capacity) {
                vertex_capacity *= 2;
                vertices = (Vec3*) realloc(vertices, vertex_capacity * sizeof(Vec3));
            }
            sscanf(line, "v %f %f %f", &vertices[pos].x, &vertices[pos].y, &vertices[pos].z);
        } else if (line[0] == 'f') {
            int pos = num_faces*3;
            num_faces++;
            if (num_faces > face_capacity) {
                face_capacity *= 2;
                faces = (int*) realloc(faces, 3 * face_capacity * sizeof(int));
            }
            sscanf(line, "f %d %d %d", &faces[pos], &faces[pos + 1], &faces[pos + 2]);
            for (int i = 0; i < 3; i++) {
                faces[pos + i] -= 1;
            }
        }
    }
    fclose(file);
    Object* object = (Object*) malloc(sizeof(Object));
    object->vertices = vertices;
    object->num_vertices = num_vertices;
    object->num_faces = num_faces;
    object->faces = faces;
    object->colors = (Color*) malloc(num_faces * sizeof(Color));
    for (int i = 0; i < num_faces; i++) {
        object->colors[i] = color;
    }
    return object;
}

void save_image(unsigned char* image, int width, int height, char* filename) {
    stbi_write_png(filename, width, height, 3, image, width * 3);
}

// gcc render.c -lm -O3 && ./a.out
int main() {
    int width = 100;
    int height = 100;

    Vec3 origin = {2.0, 10.0, 8.0};
    Vec3 destination = {0.0, 5.0, 0.0};
    Vec3 direction = vec3_sub(destination, origin);
    direction = vec3_normalize(direction);
    float fov = 90.0;

    Object** objects = (Object**) malloc(1 * sizeof(Object*));
    int num_objects = 1;

    // Object* cube = create_debug_cube(8.0);
    // objects[0] = cube;

    Object* obj = read_obj("data/bunny.obj", (Color) {255, 0, 0});
    for (int i = 0; i < obj->num_vertices; i++) {
        obj->vertices[i] = vec3_scale(obj->vertices[i], 50.0);
    }
    objects[0] = obj;

    unsigned char* image = render(width, height, objects, num_objects, origin, direction, fov);
    save_image(image, width, height, "output/c.png");
    
    // Clean up memory
    free(image);
    free(objects[0]->vertices);
    free(objects[0]->faces);
    free(objects[0]->colors);
    free(objects[0]);
    free(objects);

}

File: Render.java
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import java.awt.image.BufferedImage;
import java.io.BufferedReader;
import java.io.File;
import javax.imageio.ImageIO;

public class Render {

    public static class Vec3 {
        public double x, y, z;

        public Vec3(double x, double y, double z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public Vec3 add(Vec3 other) {
            return new Vec3(this.x + other.x, this.y + other.y, this.z + other.z);
        }

        public Vec3 sub(Vec3 other) {
            return new Vec3(this.x - other.x, this.y - other.y, this.z - other.z);
        }

        public Vec3 scale(double scalar) {
            return new Vec3(this.x * scalar, this.y * scalar, this.z * scalar);
        }

        public double dot(Vec3 other) {
            return this.x * other.x + this.y * other.y + this.z * other.z;
        }

        public double length() {
            return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
        }

        public Vec3 normalize() {
            return this.scale(1.0 / this.length());
        }

        public Vec3 cross(Vec3 other) {
            return new Vec3(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z,
                    this.x * other.y - this.y * other.x);
        }
    }

    public static class Triangle {
        public Vec3 v0, v1, v2;

        public Triangle(Vec3 v0, Vec3 v1, Vec3 v2) {
            this.v0 = v0;
            this.v1 = v1;
            this.v2 = v2;
        }
    }

    public static class Color {
        public int r, g, b;

        public Color(int r, int g, int b) {
            this.r = r;
            this.g = g;
            this.b = b;
        }
    }

    public static class Triple<T, U, V> {
        public T first;
        public U second;
        public V third;

        public Triple(T first, U second, V third) {
            this.first = first;
            this.second = second;
            this.third = third;
        }
    }

    public static class RayHit {
        public float t;
        public Vec3 normal;

        public RayHit(float t, Vec3 normal) {
            this.t = t;
            this.normal = normal;
        }
    }

    public static class Object {
        public Vec3[] vertices;
        public List<Triple<Integer, Integer, Integer>> faces;
        public Color[] colors;

        public Object(Vec3[] vertices, List<Triple<Integer, Integer, Integer>> faces, Color[] colors) {
            this.vertices = vertices;
            this.faces = faces;
            this.colors = colors;
        }
    }

    public static RayHit rayTriangleIntersection(Vec3 origin, Vec3 direction, Triangle triangle) {
        Vec3 A = triangle.v0;
        Vec3 B = triangle.v1;
        Vec3 C = triangle.v2;

        Vec3 e1 = B.sub(A);
        Vec3 e2 = C.sub(A);
        Vec3 p = direction.cross(e2);

        double det = e1.dot(p);
        if (det < 1e-6) {
            return null;
        }

        double inv_det = 1.0 / det;
        Vec3 t = origin.sub(A);
        double u = t.dot(p) * inv_det;
        if (u < 0.0 || u > 1.0) {
            return null;
        }

        Vec3 q = t.cross(e1);
        double v = direction.dot(q) * inv_det;
        if (v < 0.0 || u + v > 1.0) {
            return null;
        }

        float dist = (float) (e2.dot(q) * inv_det);
        Vec3 normal = e1.cross(e2).normalize();

        return new RayHit(dist, normal);
    }

    public static Triple<Vec3, Vec3, Vec3> calculateCameraBasis(Vec3 direction) {
        Vec3 forward = direction.normalize();
        Vec3 worldUp = new Vec3(0.0, 1.0, 0.0);
        Vec3 right = worldUp.cross(forward);
        if (right.length() < 1e-6) {
            worldUp = new Vec3(1.0, 0.0, 0.0);
            right = worldUp.cross(forward);
        }
        right = right.normalize();
        Vec3 up = forward.cross(right);
        up = up.normalize();
        up = up.scale(-1.0);
        right = right.scale(-1.0);
        return new Triple<>(forward, right, up);
    }

    public static Color[][] render(int width, int height, List<Object> objects, Vec3 origin, Vec3 direction,
            float fov) {

        Vec3 lightDir = new Vec3(0.0, 0.0, 1.0);

        Color[][] image = new Color[width][height];
        float[][] depthMap = new float[width][height];

        Triple<Vec3, Vec3, Vec3> cameraBasis = calculateCameraBasis(direction);
        Vec3 forward = cameraBasis.first;
        Vec3 right = cameraBasis.second;
        Vec3 up = cameraBasis.third;

        float aspectRatio = (float) width / height;
        float fovRad = fov * (float) Math.PI / 180.0f;
        float halfTanFov = (float) Math.tan(fovRad * 0.5f);

        for (int i = 0; i < width; i++) {
            for (int j = 0; j < height; j++) {
                float x = (2.0f * (i + 0.5f) / width - 1.0f) * aspectRatio * halfTanFov;
                float y = (2.0f * (j + 0.5f) / height - 1.0f) * halfTanFov;
                Vec3 rayDirection = forward.add(right.scale(x)).add(up.scale(y));
                rayDirection = rayDirection.normalize();

                float closestT = Float.POSITIVE_INFINITY;
                Color closestColor = new Color(0, 0, 0);

                for (Object object : objects) {
                    for (int faceIndex = 0; faceIndex < object.faces.size(); faceIndex++) {
                        Triple<Integer, Integer, Integer> face = object.faces.get(faceIndex);
                        int idx0 = face.first;
                        int idx1 = face.second;
                        int idx2 = face.third;

                        Vec3 a = object.vertices[idx0];
                        Vec3 b = object.vertices[idx1];
                        Vec3 c = object.vertices[idx2];

                        Triangle triangle = new Triangle(a, b, c);

                        RayHit hit = rayTriangleIntersection(origin, rayDirection, triangle);
                        if (hit != null && hit.t > 0.0f && hit.t < closestT) {
                            closestT = hit.t;
                            Color color = object.colors[faceIndex];
                            double d = Math.max(0.4, hit.normal.dot(lightDir));
                            closestColor = new Color(
                                (int) (color.r * d),
                                (int) (color.g * d),
                                (int) (color.b * d)
                            );
                        }
                    }
                }

                image[i][j] = closestColor;
                depthMap[i][j] = closestT;
            }
        }

        return image;
    }

    public static void writePPM(Color[][] image, String filename) {
        try (FileOutputStream fos = new FileOutputStream(filename)) {
            fos.write(("P3\n" + image.length + " " + image[0].length + "\n255\n").getBytes());
            int width = image.length;
            int height = image[0].length;
            for (int j = 0; j < height; j++) {
                for (int i = 0; i < width; i++) {
                    fos.write((image[i][j].r + " " + image[i][j].g + " " + image[i][j].b + "\n").getBytes());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void saveImage(Color[][] image, String filename) {
        BufferedImage img = new BufferedImage(image.length, image[0].length, BufferedImage.TYPE_INT_RGB);
        for (int i = 0; i < image.length; i++) {
            for (int j = 0; j < image[i].length; j++) {
                img.setRGB(i, j, image[i][j].r << 16 | image[i][j].g << 8 | image[i][j].b);
            }
        }
        try {
            ImageIO.write(img, "png", new File(filename));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static Object createDebugCube(float size) {
        float halfSize = size / 2.0f;
        Vec3[] vertices = new Vec3[8];
        vertices[0] = new Vec3(-halfSize, -halfSize, -halfSize);
        vertices[1] = new Vec3( halfSize, -halfSize, -halfSize);
        vertices[2] = new Vec3( halfSize,  halfSize, -halfSize);
        vertices[3] = new Vec3(-halfSize,  halfSize, -halfSize);
        vertices[4] = new Vec3(-halfSize, -halfSize,  halfSize);
        vertices[5] = new Vec3( halfSize, -halfSize,  halfSize);
        vertices[6] = new Vec3( halfSize,  halfSize,  halfSize);
        vertices[7] = new Vec3(-halfSize,  halfSize,  halfSize);

        List<Triple<Integer, Integer, Integer>> faces = new ArrayList<>();
        faces.add(new Triple<>(0, 2, 1));
        faces.add(new Triple<>(0, 3, 2));

        faces.add(new Triple<>(4, 5, 6));
        faces.add(new Triple<>(4, 6, 7));

        faces.add(new Triple<>(0, 4, 7));
        faces.add(new Triple<>(0, 7, 3));

        faces.add(new Triple<>(1, 2, 6));
        faces.add(new Triple<>(1, 6, 5));

        faces.add(new Triple<>(0, 1, 5));
        faces.add(new Triple<>(0, 5, 4));

        faces.add(new Triple<>(2, 3, 7));
        faces.add(new Triple<>(2, 7, 6));

        Color[] colors = new Color[12];
        colors[0] = new Color(0, 0, 255);
        colors[1] = new Color(0, 0, 255);
        colors[2] = new Color(0, 255, 0);
        colors[3] = new Color(0, 255, 0);
        colors[4] = new Color(255, 0, 0);
        colors[5] = new Color(255, 0, 0);
        colors[6] = new Color(255, 255, 0);
        colors[7] = new Color(255, 255, 0);
        colors[8] = new Color(255, 0, 255);
        colors[9] = new Color(255, 0, 255);
        colors[10] = new Color(0, 255, 255);
        colors[11] = new Color(0, 255, 255);

        return new Object(vertices, faces, colors);
    }

    public static Object readObj(String filename, Color color) {
        List<Vec3> vertices = new ArrayList<>();
        List<Triple<Integer, Integer, Integer>> faces = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.startsWith("v ")) {
                    vertices.add(new Vec3(Float.parseFloat(line.split(" ")[1]), Float.parseFloat(line.split(" ")[2]), Float.parseFloat(line.split(" ")[3])));
                }
                if (line.startsWith("f ")) {
                    String[] indices = line.split(" ");
                    int[] intIndices = new int[indices.length];
                    for(int i = 1; i < indices.length; i++) {
                        intIndices[i] = Integer.parseInt(indices[i].split("/")[0]) - 1;
                    }
                    faces.add(new Triple<>(intIndices[1], intIndices[2], intIndices[3]));
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        Color[] colors = new Color[faces.size()];
        for (int i = 0; i < faces.size(); i++) {
            colors[i] = color;
        }

        return new Object(vertices.toArray(new Vec3[0]), faces, colors);
    }

    public static void main(String[] args) {

        int width = 100;
        int height = 100;

        Vec3 origin = new Vec3(2.0f, 10.0f, 8.0f);
        Vec3 destination = new Vec3(0.0f, 5.0f, 0.0f);
        Vec3 direction = destination.sub(origin).normalize();
        float fov = 90.0f;

        ArrayList<Object> objects = new ArrayList<>();
        // objects.add(createDebugCube(8.0f));
        Object bunny = readObj("data/bunny.obj", new Color(255, 0, 0));
        bunny.vertices = Arrays.stream(bunny.vertices).map(v -> v.scale(50.0f)).toArray(Vec3[]::new);
        objects.add(bunny);

        Color[][] image = render(width, height, objects, origin, direction, fov);
        // writePPM(image, "output/java.ppm");
        saveImage(image, "output/java.png");
    }
}
